### ISSUE TITLE
dnsdist: OCSP Stapling with local PKI doesn't reloadAllCertificates

### ISSUE THREAD

**Issue by ancarpan** [PowerDNS/pdns#8431]
Hello,
I'm testing OCSP Stapling with local PKI and I've found the following issue:

Current dnsdist.cfg has:

```
generateOCSPResponse("/etc/dnsdist/pki/ECDSA/ServerCert/ServerCert_Cert.pem",
                    "/etc/dnsdist/pki/ECDSA/IssuingCA/Issuing_CA_Cert.pem",
                    "/etc/dnsdist/pki/ECDSA/IssuingCA/Issuing_CA_Private.key",
                    "/var/run/dnsdist/ssl/ECDSA_OCSP.pem",
                    10, 10)
generateOCSPResponse("/etc/dnsdist/pki/RSA/ServerCert/ServerCert_Cert.pem",
                    "/etc/dnsdist/pki/RSA/IssuingCA/Issuing_CA_Cert.pem",
                    "/etc/dnsdist/pki/RSA/IssuingCA/Issuing_CA_Private.key",
                    "/var/run/dnsdist/ssl/RSA_OCSP.pem",
                    10, 10)

-- add the DNS-over-HTTPs receiver threads for IPv4
addDOHLocal("0.0.0.0:443",
          { "/etc/dnsdist/pki/ECDSA/ECDSA_chain.pem","/etc/dnsdist/pki/RSA/RSA_chain.pem" },
          { "/etc/dnsdist/pki/ECDSA/ServerCert/ServerCert_Private.key","/etc/dnsdist/pki/RSA/ServerCert/ServerCert_Private.key" },
          {"/"}, {reusePort=true, ocspResponses={"/var/run/dnsdist/ssl/ECDSA_OCSP.pem", "/var/run/dnsdist/ssl/RSA_OCSP.pem"}})
addDOHLocal("0.0.0.0:443",
          { "/etc/dnsdist/pki/ECDSA/ECDSA_chain.pem","/etc/dnsdist/pki/RSA/RSA_chain.pem" },
          { "/etc/dnsdist/pki/ECDSA/ServerCert/ServerCert_Private.key","/etc/dnsdist/pki/RSA/ServerCert/ServerCert_Private.key" },
          {"/"}, {reusePort=true, ocspResponses={"/var/run/dnsdist/ssl/ECDSA_OCSP.pem", "/var/run/dnsdist/ssl/RSA_OCSP.pem"}})
addDOHLocal("0.0.0.0:443",
          { "/etc/dnsdist/pki/ECDSA/ECDSA_chain.pem","/etc/dnsdist/pki/RSA/RSA_chain.pem" },
          { "/etc/dnsdist/pki/ECDSA/ServerCert/ServerCert_Private.key","/etc/dnsdist/pki/RSA/ServerCert/ServerCert_Private.key" },
          {"/"}, {reusePort=true, ocspResponses={"/var/run/dnsdist/ssl/ECDSA_OCSP.pem", "/var/run/dnsdist/ssl/RSA_OCSP.pem"}})
```
At startup I can see DER files produced with correct data:
```
[root@lapp-dohc01dns ~]# openssl ocsp -respin /var/run/dnsdist/ssl/ECDSA_OCSP.pem -text|grep Update
    This Update: Oct 18 09:11:22 2019 GMT
    Next Update: Oct 28 09:21:22 2019 GMT
```
and testing via TCP socket gives consistent info:
```
$ echo ""| openssl s_client -cipher RSA -connect 213.1.145.52:443 -status -servername test01.doh.dns.networks.bt.com 2>&1 | grep Update
    This Update: Oct 18 09:11:22 2019 GMT
    Next Update: Oct 28 09:21:22 2019 GMT
```

I then trigger a reloadAllCertificates().
```
[root@lapp-dohc01dns ~]# dnsdist -e 'reloadAllCertificates()'
access to dnsdist.REFUSED is deprecated, please use DNSRCode.REFUSED
access to dnsdist.ANY is deprecated, please use DNSQType.ANY
access to dnsdist.NXDOMAIN is deprecated, please use DNSRCode.NXDOMAIN
access to dnsdist.SERVFAIL is deprecated, please use DNSRCode.SERVFAIL
```
The files on disk are regenerated with current timestamp 
```
[root@lapp-dohc01dns ~]# ls -la /var/run/dnsdist/ssl/ECDSA_OCSP.pem
-rw-r--r--. 1 root root 535 Oct 18 10:32 /var/run/dnsdist/ssl/ECDSA_OCSP.pem
[root@lapp-dohc01dns ~]# openssl ocsp -respin /var/run/dnsdist/ssl/ECDSA_OCSP.pem -text|grep Update
    This Update: Oct 18 09:32:06 2019 GMT
    Next Update: Oct 28 09:42:06 2019 GMT
```
But "openssl s_client" still has previous OCSP data:
```
$ echo ""| openssl s_client -cipher RSA -connect 213.1.145.52:443 -status -servername test01.doh.dns.networks.bt.com 2>&1 | grep Update
    This Update: Oct 18 09:11:22 2019 GMT
    Next Update: Oct 28 09:21:22 2019 GMT
```

Info:
```
[root@lapp-dohc01dns ~]# dnsdist -V
dnsdist 1.4.0-rc3.64.master.gfdf831b1a (Lua 5.1.4 [LuaJIT 2.0.4])
Enabled features: cdb dns-over-tls(gnutls openssl) dns-over-https(DOH) dnscrypt fstrm ipcipher libsodium lmdb protobuf re2 recvmmsg/sendmmsg snmp systemd
```

**Comment by rgacogne** [comment]
> I then trigger a reloadAllCertificates().

Could you check is there is anything relevant in the log (syslog or journalctl)? It might be a permission issue since we drop privileges after startup.

> The files on disk are regenerated with current timestamp

Did you re-generate them by calling `generateOCSPResponse()` directly? Because `reloadAllCertificates()` is not expected to do that so I'm a bit surprised.

I did a quick test by calling `generateOCSPResponse()` via the console to generate a new OCSP file, then issued a `reloadAllCertificates()` and I did get the new OCSP response, but I probably have something different in my setup.

**Comment by ancarpan** [comment]
> > I then trigger a reloadAllCertificates().
> Could you check is there is anything relevant in the log (syslog or journalctl)? It might be a permission issue since we drop privileges after startup.

Yes something in the logs: sorry for not finding this earlier:
```
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: Got control connection from 127.0.0.1:55772
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:0200100D:system library:fopen:Permission denied:bss_file.c:402:fopen('/etc/dnsdist/pki/ECDSA/ECDSA_chain.pem','r')
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:20074002:BIO routines:FILE_CTRL:system lib:bss_file.c:404:
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:140DC002:SSL routines:SSL_CTX_use_certificate_chain_file:system lib:ssl_rsa.c:701:
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: Error reloading certificates for frontend 0.0.0.0:443: Failed to setup SSL/TLS for DoH listener, an error occurred while trying to load the DOH server certificate file: /etc/dnsdist/pki/ECDSA/ECDSA_chain.pem
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:0200100D:system library:fopen:Permission denied:bss_file.c:402:fopen('/etc/dnsdist/pki/ECDSA/ECDSA_chain.pem','r')
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:20074002:BIO routines:FILE_CTRL:system lib:bss_file.c:404:
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:140DC002:SSL routines:SSL_CTX_use_certificate_chain_file:system lib:ssl_rsa.c:701:
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: Error reloading certificates for frontend 0.0.0.0:443: Failed to setup SSL/TLS for DoH listener, an error occurred while trying to load the DOH server certificate file: /etc/dnsdist/pki/ECDSA/ECDSA_chain.pem
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:0200100D:system library:fopen:Permission denied:bss_file.c:402:fopen('/etc/dnsdist/pki/ECDSA/ECDSA_chain.pem','r')
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:20074002:BIO routines:FILE_CTRL:system lib:bss_file.c:404:
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: 140200239802112:error:140DC002:SSL routines:SSL_CTX_use_certificate_chain_file:system lib:ssl_rsa.c:701:
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: Error reloading certificates for frontend 0.0.0.0:443: Failed to setup SSL/TLS for DoH listener, an error occurred while trying to load the DOH server certificate file: /etc/dnsdist/pki/ECDSA/ECDSA_chain.pem
Oct 21 13:58:54 lapp-dohc01dns dnsdist[24499]: Closed control connection from 127.0.0.1:55772
```

> > The files on disk are regenerated with current timestamp
> Did you re-generate them by calling `generateOCSPResponse()` directly? Because `reloadAllCertificates()` is not expected to do that so I'm a bit surprised.

No, since my tests with  `generateOCSPResponse()` were failing (see our mail thread about that), I just tried `reloadAllCertificates()`. I was quite surprised too.

> I did a quick test by calling `generateOCSPResponse()` via the console to generate a new OCSP file, then issued a `reloadAllCertificates()` and I did get the new OCSP response, but I probably have something different in my setup.

How did you manage to have `generateOCSPResponse()` working? You removed the root files by hand?

**Comment by rgacogne** [comment]
> Yes something in the logs: sorry for not finding this earlier:

Right, so we have a permission issue, the user dnsdist is running under is not allowed to read the certificate files. 
> No, since my tests with `generateOCSPResponse()` were failing (see our mail thread about that), I just tried `reloadAllCertificates()`. I was quite surprised too.

Then I'm afraid I don't know what happened, as `reloadAllCertificates()` will clearly not renew the OCSP file :-/

> How did you manage to have `generateOCSPResponse()` working? You removed the root files by hand?

No, it's because on my setup the dnsdist process is started as dnsdist, using something similar to what we will do once https://github.com/PowerDNS/pdns/pull/7820 has been merged.

**Comment by ancarpan** [comment]
> > No, since my tests with `generateOCSPResponse()` were failing (see our mail thread about that), I just tried `reloadAllCertificates()`. I was quite surprised too.
> Then I'm afraid I don't know what happened, as `reloadAllCertificates()` will clearly not renew the OCSP file :-/

AndreaT suggested the following: `dnsdist -e' wil execute anyway dnsdist.cfg. This is probably what is happening.

I think the original issue is related to permissions and misunderstanding of how `-e` works. We can probably close this bug.

**Comment by rgacogne** [comment]
That's actually a bug, `generateOCSPResponse()` should not be executed in client or command mode! I'll see if I can fix that!

**Comment by rgacogne** [comment]
Requires #8456 for a proper fix.

**Comment by rgacogne** [comment]
I believe this issue has been fixed by #8456, closing.
